internal/controllers/machine/machine_controller.go

```diff
@@ -791,38 +805,89 @@ func shouldRequeueDrain(now time.Time, lastDrain time.Time) (time.Duration, bool
 		// of unreachable nodes when a volume is attached.
 		log.Info("Node is unreachable, skip waiting for volume detachment.")
-		return false, nil
+		return ctrl.Result{}, nil
 	}
 
-	if len(node.Status.VolumesAttached) != 0 {
-		log.Info("Waiting for Node volumes to be detached")
-		return true, nil
+	// Check waitForVolumeDetach cache to ensure we won't retry before waitForVolumeDetachInterval.
+	// Note: This is intentionally only done if we have Pods to evict, because otherwise
+	// even the "no-op" drain would be requeued.
+	if cacheEntry, ok := r.waitForVolumeDetachCache.Has(client.ObjectKeyFromObject(machine)); ok {
+		if requeueAfter, requeue := shouldRequeue(time.Now(), cacheEntry.LastProceed, waitForVolumeDetachInterval); requeue {
+			log.Info(fmt.Sprintf("Requeuing in %s because there already was a wiat for node volumes in the last %s", requeueAfter, waitForVolumeDetachInterval))
+			return ctrl.Result{RequeueAfter: requeueAfter}, nil
+		}
 	}
 
-	return false, nil
+	// Get two sets of information about volumes currently attached to the node:
+	// * VolumesAttached names from node.Status.VolumesAttached
+	// * PersistentVolume names from VolumeAttachments with status.Attached set to true
+	attachedVolumeNames, attachedPVNames, err := getAttachedVolumeInformation(ctx, remoteClient, node)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Return early if there are no volumes to wait for getting detached.
+	if len(attachedVolumeNames) == 0 && len(attachedPVNames) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Add entry an to the cache so we won't retry before waitForVolumeDetachInterval.
+	r.waitForVolumeDetachCache.Add(drain.CacheEntry{
+		Machine:     client.ObjectKeyFromObject(machine),
+		LastProceed: time.Now(),
+	})
+
+	// Get all PVCs we want to ignore because they have pods running which got skipped during drain.
+	pvcsToIgnoreFromPods, err := getPersistentVolumeClaimsToIgnore(ctx, remoteClient, nodeName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// List all PersistentVolumes and return the ones we want to wait for.
+	pvsWaitingForDetach, err := getPersistentVolumesWaitingForDetach(ctx, remoteClient, attachedVolumeNames, attachedPVNames, pvcsToIgnoreFromPods)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If no pvs were found we have to wait for and there are no unmatched VolumeHandles
+	// or PVs from VolumeAttachments, then we are finished with waiting for volumes.
+	if len(pvsWaitingForDetach) == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// Otherwise log the referred PersistentVolumeClaims (max 5) we are still waiting for to be detached.
+	// Sort the slice before logging to be deterministic.
+	slices.SortFunc(pvsWaitingForDetach, func(a *corev1.PersistentVolume, b *corev1.PersistentVolume) int {
+		return cmp.Compare(a.GetName(), b.GetName())
+	})
+	log.Info("Waiting for Node volumes to be detached",
+		"PersistentVolumeClaims", clog.ObjNamesString(pvsWaitingForDetach),
```

Comment:
Wondering if we can provide the info about which pod the volumes belong to, but this can be also a follow up iteration

Reply:
With the current code: this would require us to map all pods from the drainer to match them accordingly (we already list them in the drainer so this won't be additional list calls). Normally these pods should already be gone because we wait for this when draining, so we won't get any further information.

The only exception to this is when drain is done due to hitting the timeout.

I'm okay to do this as follow-up, but important to note that we may not always gather more information.

Reply:
> but important to note that we may not always gather more information.

I think that's the key point.
Let's ignore for now